File: home/ubuntu/remitflow/54remitflow-unified-platform/android-native/app/src/main/java/com/remittance/viewmodels/LivenessViewModel.kt
package com.remittance.viewmodels

import android.content.Context
import android.graphics.Bitmap
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.remittance.networking.LivenessApiService
import com.remittance.networking.RetrofitClient
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.MultipartBody
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.ByteArrayOutputStream

sealed class LivenessState {
    object Idle : LivenessState()
    object Loading : LivenessState()
    data class Success(val isLive: Boolean, val confidence: Float) : LivenessState()
    data class Error(val message: String) : LivenessState()
}

class LivenessViewModel : ViewModel() {

    private val _livenessState = MutableStateFlow<LivenessState>(LivenessState.Idle)
    val livenessState: StateFlow<LivenessState> = _livenessState

    private val livenessApiService: LivenessApiService by lazy {
        RetrofitClient.livenessInstance.create(LivenessApiService::class.java)
    }

    fun checkLiveness(context: Context, selfieBitmap: Bitmap) {
        viewModelScope.launch {
            _livenessState.value = LivenessState.Loading
            try {
                val stream = ByteArrayOutputStream()
                selfieBitmap.compress(Bitmap.CompressFormat.JPEG, 100, stream)
                val byteArray = stream.toByteArray()

                val requestFile = byteArray.toRequestBody("image/jpeg".toMediaTypeOrNull(), 0, byteArray.size)
                val body = MultipartBody.Part.createFormData("selfie_image", "selfie.jpg", requestFile)

                val response = livenessApiService.checkLiveness(body)

                if (response.is_live) {
                    _livenessState.value = LivenessState.Success(response.is_live, response.confidence_score)
                } else {
                    _livenessState.value = LivenessState.Error("Liveness check failed. Please try again.")
                }
            } catch (e: Exception) {
                _livenessState.value = LivenessState.Error(e.message ?: "An unknown error occurred.")
            }
        }
    }
}

File: 54remitflow-unified-platform/ios-native/RemittanceApp/ViewModels/LivenessViewModel.swift
import SwiftUI
import Combine

@MainActor
class LivenessViewModel: ObservableObject {
    @Published var livenessState: LivenessState = .idle

    private var apiClient = APIClient.shared
    private var cancellables = Set<AnyCancellable>()

    func checkLiveness(selfieImage: UIImage) {
        livenessState = .loading

        guard let imageData = selfieImage.jpegData(compressionQuality: 0.8) else {
            livenessState = .error("Failed to convert image to data.")
            return
        }

        Task {
            do {
                let result: LivenessResult = try await apiClient.upload(
                    .checkLiveness,
                    data: imageData,
                    fileName: "selfie.jpg",
                    mimeType: "image/jpeg"
                )

                if result.is_live {
                    livenessState = .success(isLive: result.is_live, confidence: result.confidence_score)
                } else {
                    livenessState = .error("Liveness check failed. Please try again.")
                }
            } catch {
                livenessState = .error(error.localizedDescription)
            }
        }
    }
}

enum LivenessState {
    case idle
    case loading
    case success(isLive: Bool, confidence: Float)
    case error(String)
}
